rfcs/text/0000_alert_actions_service.md

```diff
@@ -0,0 +1,210 @@
+- Start Date: 2019-02-22
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+An actions service for registering *action instances* for a specific
+*action type* for a given *connector*.
+
+# Basic example
+
+An example of a preconfigured `instance` of a `actionType` with a specific
+`connectorType` (i.e. slack).
+
+```JS
+server.actions.fire({
+  action: 'send message to slack',
+  actionType: 'send message',
+  params: {
+    message: 'This is a test message from kibana actions service',
+    title: 'custom title',
+  },
+})
+```
+
+# Motivation
+
+In order to support multiple integrations with external services such as slack,
+email, pager-duty, twitter, etc. There is a need for an interface that an alert
+task would use to execute actions with customized content that is specific to
+that alert. In order to increase customization and extensibility there is a need
+for a way the task could fire an action without knowing the specifics of a
+particular service (i.e email, slack)
+
+# Detailed design
+
+This proposal defines one method to allow an alert task to be decoupled from the
+action(s) that the user wants the alert task to execute when the alert is
+*triggered*. The `action` is a preconfigured saved `instance` of a `actionType`
+for a specific `connectorType`
+
+## Executing Actions
+
+Each action instance can be fired with optional action type parameter. You need
+to have a defined action instance in order to fire an action.
+
+```JS
+server.actions.fire({
+  action: 'send message to slack',
+  actionType: 'notification',
+  params: {
+    message: 'This is a test message from kibana actions service',
+    title: 'custom title',
+    // destination: '<slack url>' // optional
+  },
+})
+```
+
+The action `send message to slack` is the instance that is being fired. See
+below for details how to create them
+
+## Creating Instances
+
+Instances are saved in the Kibana index and are used in order to `fire` an action with parameters. The
+`params` required are specific to the `actionType`. Changing the `actionType` also
+changes what `connectorTypes` are available for it.
+
+
+```JS
+server.actions.instance({
+  name: 'send message to slack',
+  actionType: 'notification',
+  connectorType: 'slack',
+  params: {
+    destination: '<slack url>',
+  },
+  connectorParams: {
+    channel: '#bot-playground',
+  },
+});
+```
+
+*Note: instances are saved objects that are space aware*
+
+## The Connectors
+
+The `connector` is what defines what happens when an action is *fired*. This
+will define the particular integration such as slack, email, etc.
+
+```JS
+server.actions.registerConnector({
+  actionType: 'notification',
+  connectorType: 'slack',
```

Comment:
Should connectors/handlers define param types as well?

Reply:
Absolutely, the connectorParam `channel`'s type should be defined here! And the instance method should enforce proper creation of the param structure and types.

This might be different when using typescript though as you could probably just specify the connectorParams type and infer the type from that.
```JS
connectorParams: {
  { name: 'channel', type: 'string' }
},
```